library/src/main/java/com/chuckerteam/chucker/internal/ui/transaction/TransactionPayloadFragment.kt

```diff
@@ -71,21 +71,50 @@ internal class TransactionPayloadFragment :
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+
+        val result = mutableListOf<TransactionPayloadItem>()
+        val adapter = TransactionBodyAdapter(result)
+        payloadBinding.responseRecyclerView.adapter = adapter
+        payloadBinding.responseRecyclerView.setHasFixedSize(true)
+
+        viewModel.encodeUrl.observe(
+            viewLifecycleOwner,
+            Observer { encode ->
+                if (adapter.itemCount == 0) {
+                    return@Observer
+                }
+                val transaction = viewModel.transaction.value ?: return@Observer
+                uiScope.launch {
+                    updateUpdater(result, adapter, transaction, !encode)
+                }
+            }
+        )
+
         viewModel.transaction.observe(
             viewLifecycleOwner,
             Observer { transaction ->
                 if (transaction == null) return@Observer
                 uiScope.launch {
                     showProgress()
-                    val result = processPayload(type, transaction)
-                    payloadBinding.responseRecyclerView.adapter = TransactionBodyAdapter(result)
-                    payloadBinding.responseRecyclerView.setHasFixedSize(true)
+                    val encode = viewModel.encodeUrl.value ?: true
+                    updateUpdater(result, adapter, transaction, !encode)
                     hideProgress()
                 }
             }
         )
     }
 
+    private suspend fun updateUpdater(
+        data: MutableList<TransactionPayloadItem>,
+        adapter: TransactionBodyAdapter,
+        transaction: HttpTransaction,
+        formatted: Boolean
+    ) {
+        data.clear()
```

Comment:
In general this function shouldn't be in this fragment. It should be inside the adapter itself named something like `setItems()`.

Reply:
Also, in case of having such thing as `setItems()` function you can remove adapter initialisation with an empty list and make its declaration lazy for cases when we have empty payload.

Reply:
Definitely, I didn't like this to begin with and I guess I just didn't want to change too many things. I'll fix this. But I'm not sure I get exactly what you meant with your proposed solution. Are you suggesting the use of a lazy property or something else?

Here I can only see a nullable list as marginally better, given that the moment you attach the adapter to the RecyclerView its `getItemCount()` gets called, so anything that creates the list lazily would do it immediately.

Reply:
Yes, I suggest to have a lazy property and create it only when we have some result to show.
However, now I think it might be not the best idea, since it will cause logcat show `No adapter attached`, like it does now.